src/lib.rs

```diff
@@ -1,123 +1,289 @@
 
-        let sgx_signing_cert_pem = &parse_pem(include_bytes!("../data/signing_cert.pem")).unwrap()[0];
-        let sgx_signing_cert = parse_x509_der(&sgx_signing_cert_pem.contents);
+    // Verify that the CRL is signed by Intel and add it to the store.
+    trust_store
+        .push_unverified_crl(collateral.root_ca_crl.clone())
+        .context("failed to verify root ca crl")?;
 
-        assert!(validate_tcbinfov2(&tcbinfov2, &sgx_signing_cert, PINNED_TIME));
+    // Verify PCK CRL Chain and add it to the store.
+    let pck_issuer = trust_store
+        .verify_chain_leaf(&collateral.pck_crl_issuer_chain)
+        .context("failed to verify pck crl issuer chain")?;
+
+    // Verify the pck crl and add it to the store.
+    pck_issuer
+        .pk
+        .verify(&collateral.pck_crl)
+        .map_err(|e| anyhow::anyhow!("failed to verify pck crl signature: {}", e))?;
+    if !collateral.pck_crl.valid_at(current_time) {
+        bail!("expired pck crl");
     }
+    trust_store.push_trusted_crl(collateral.pck_crl.clone());
 
-    #[test]
-    fn test_quotev4() {
-        let quotev4_slice = include_bytes!("../data/quote_tdx_00806f050000.dat");
-        let quotev4 = QuoteV4::from_bytes(quotev4_slice);
-        assert_eq!(quotev4.header.version, 4);
+    // Verify TCB Info Issuer Chain
+    let tcb_issuer = trust_store
+        .verify_chain_leaf(&collateral.tcb_info_issuer_chain)
+        .context("failed to verify tcb info issuer chain")?;
+
+    // Get TCB Signer Public Key
+    let tcb_signer = tcb_issuer
+        .cert
+        .tbs_certificate
+        .subject_public_key_info
+        .subject_public_key
+        .as_bytes()
+        .context("missing tcb signer public key")?;
+
+    // We are making big assumption here that the key is ECDSA P-256
+    let tcb_signer = p256::ecdsa::VerifyingKey::from_sec1_bytes(tcb_signer)
+        .context("invalid tcb signer public key")?;
+
+    // Verify the TCB Info
+    let tcb_info = collateral
+        .tcb_info
+        .as_tcb_info_and_verify(tcb_signer)
+        .context("failed to verify tcb info signature")?;
+
+    // Verify the quote's pck signing certificate chain
+    let _pck_signer = trust_store
+        .verify_chain_leaf(&quote.signature.pck_cert_chain)
+        .context("failed to verify quote support pck signing certificate chain")?;
+
+    // Verify the quote identity issuer chain
+    let _qe_id_issuer = trust_store
+        .verify_chain_leaf(&collateral.qe_identity_issuer_chain)
+        .context("failed to verify pck crl issuer certificate chain")?;
+
+    Ok(tcb_info)
+}
+
+fn verify_quote(collateral: &Collateral, quote: &Quote) -> anyhow::Result<()> {
+    verify_quote_enclave_source(collateral, quote)?;
+    verify_quote_signatures(quote)?;
+    Ok(())
+}
+
+/// Verify the quote enclave source
+fn verify_quote_enclave_source(collateral: &Collateral, quote: &Quote) -> anyhow::Result<()> {
```

Comment:
`qe_identity` expiration check is missing

Reply:
Done